### ISSUE TITLE
TC004 probably shouldn't trigger on typing-modules

### ISSUE THREAD

**Issue by mikeshardmind** [astral-sh/ruff#16045]
### Description

relevant parts of config:

```toml
[tool.ruff]

line-length = 90
target-version = "py312"
preview = true

[tool.ruff.format]
line-ending = "lf"

[tool.ruff.lint]
select = [
    "A", "ANN", "ASYNC", "B", "BLE", "C4", "COM", "D", "DOC", "DTZ", "E",
    "EM", "ERA", "F", "FA", "FBT", "FURB", "G", "I", "INP", "ISC", "NPY",
    "PD", "PERF", "PGH", "PIE", "PLC", "PLE", "PLR", "PLW", "PTH", "PYI",
    "Q", "Q003", "RET", "RSE", "RUF", "S", "SIM", "SLOT", "T20", "TC", "TID",
    "TRY", "UP", "YTT"
]

typing-modules = ["async_utils._typings"]
```

contents of `async_utils._typings`

```py
# License and context of encountering this issue available here
# https://github.com/mikeshardmind/async-utils/blob/c84b9a71f8caf5a5dcfafc675eff7dec60bd4162/src/async_utils/_typings.py#L27-L44
from __future__ import annotations

TYPE_CHECKING = False
if TYPE_CHECKING:
    from typing import Any, Literal, Never, Self
else:

    def __getattr__(name: str):
        if name in {"Any", "Literal", "Never", "Self"}:
            import typing

            return getattr(typing, name)

        msg = f"module {__name__!r} has no attribute {name!r}"
        raise AttributeError(msg)


__all__ = ["TYPE_CHECKING", "Any", "Literal", "Never", "Self"]
```

errors:

```
Error: src/async_utils/_typings.py:31:24: TC004 Move import `typing.Any` out of type-checking block. Import is used for more than type hinting.
Error: src/async_utils/_typings.py:31:29: TC004 Move import `typing.Literal` out of type-checking block. Import is used for more than type hinting.
Error: src/async_utils/_typings.py:31:38: TC004 Move import `typing.Never` out of type-checking block. Import is used for more than type hinting.
Error: src/async_utils/_typings.py:31:45: TC004 Move import `typing.Self` out of type-checking block. Import is used for more than type hinting.
```

Not sure there's a good way to handle this, this is intentionally slightly dynamic use, as throughout the rest of the module, use of typing is done with something like:

```py
from __future__ import annotations  # remove this on 3.14+

from . import _typings as t

def errors() -> t.Never:
    ...
```

Which preserves runtime introspectability, but can result in typing never being imported if that introspection capability is never used.

Checking that a corresponding use exists in else block here might be more trouble than it is worth, and it may be better to just exempt any file marked as part of `typing-modules` from this. For now I'm going to disable the rule.

**Comment by ntBre** [comment]
Thanks for the report! This looks like a change/regression in 0.9.5 (possibly related to #15719?) I was having trouble reproducing it locally but realized my global ruff install was on 0.9.3 😆 

I can actually reproduce with just `--select TC004` and `--preview` on 0.9.5:

```shell
 cat <<'EOF' | uvx ruff@0.9.5 check --select TC004 --no-cache --preview --isolated -
from __future__ import annotations

TYPE_CHECKING = False
if TYPE_CHECKING:
    from typing import Any, Literal, Never, Self
else:

    def __getattr__(name: str):
        if name in {"Any", "Literal", "Never", "Self"}:
            import typing

            return getattr(typing, name)

        msg = f"module {__name__!r} has no attribute {name!r}"
        raise AttributeError(msg)


__all__ = ["TYPE_CHECKING", "Any", "Literal", "Never", "Self"]
EOF
```

@AlexWaygood might have more insight into the intended or desired behavior here.

**Comment by AlexWaygood** [comment]
> This is probably a dumb question but is there a specific reason why the typing module uses `.py` over `.pyi`?

The idea of the typing module here is to reduce the import time of the app or library at runtime by only lazily importing the `typing` module "on demand". On more recent Python versions, the typing module is actually faster to import than lots of other popular stdlib modules such as `inspect` or `dataclasses`, but it has a historical reputation as being very slow to import, so this kind of thing is reasonably common. (It's also one of the easier modules to do this kind of thing with; it's harder to use this kind of shim module to delay the import of `dataclasses`, for example.)

**Comment by Daverball** [comment]
The main culprit here is `__all__`, it creates a runtime reference to these symbols and the `else` clause is too dynamic for ruff to understand that these symbols are actually available at runtime.

This is such a specialized case, I think it makes more sense to just `noqa: TCH004` the imports, anything else is a hack at best.

**Comment by Daverball** [comment]
Another solution is to disable `TC004` if there is a module-level `__getattr__`, since realistically we can't know at that point if a symbol is available at runtime or not.

**Comment by AlexWaygood** [comment]
> Another solution is to disable `TC004` if there is a module-level `__getattr__` and the only runtime reference is in `__all__`, since realistically we can't know at that point if a symbol is available at runtime or not.

Oh, that seems like a reasonable improvement to me!
